### PR TITLE
Implement new T&C solution on Android

### PR DESCRIPTION
**Pull request by czerwix**
## Purpose

Implement new T&C solution on Android detailed description can be found in issue [CBW-911](https://concordium.atlassian.net/browse/CBW-911).

## Changes

Added new Onboarding Activity with jetpack navigation and KoinDI. introduced new TermsFragment that will handle terms accepted by the client. added a number of todos to keep track of further work required.

Below sample usage of new changes introduced:
[concordium_terms_initial.webm](https://user-images.githubusercontent.com/11137601/235371977-c733eedd-0943-4790-b133-37c94624e43e.webm)
[concordium_terms_update.webm](https://user-images.githubusercontent.com/11137601/235371982-c633ca8c-3b7c-4ea8-bcf7-bf20cd98f87a.webm)

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above-linked CLA.

[CBW-911]: https://concordium.atlassian.net/browse/CBW-911?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -376,4 +382,18 @@ dependencies {
     implementation("org.jetbrains.kotlin:kotlin-reflect:1.8.10")
 
     implementation "androidx.security:security-crypto:1.0.0"
+
+    //DI
```
Comment (bisgardo) [inline app/build.gradle:386]: Please be more specific. And add space after `//` :)

Same below. A sentence or two about why we're depending on this/where we're using it would be great.

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -33,21 +33,37 @@
...
+            android:exported="true"
+            android:launchMode="singleTop"
+            android:screenOrientation="portrait">
+<!--            <intent-filter>-->
+<!--                <action android:name="android.intent.action.MAIN" />-->
+<!--                <category android:name="android.intent.category.LAUNCHER" />-->
+<!--            </intent-filter>-->
+        </activity>
+
+        <!--        old code-->
```
Comment (bisgardo) [inline app/src/main/AndroidManifest.xml:49]: This doesn't look done?

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -33,21 +33,37 @@
...
-            android:exported="true">
+            android:screenOrientation="portrait"
+            android:theme="@style/AppTheme.NoActionBar.Launcher">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
             <intent-filter>
                 <action android:name="android.intent.action.VIEW" />
+
```
Comment (bisgardo) [inline app/src/main/AndroidManifest.xml:63]: why these extra lines?

File: app/src/main/java/com/concordium/wallet/App.kt [hunk 4]
```diff
@@ -1,30 +1,40 @@
...
+import org.koin.android.ext.koin.androidLogger
+import org.koin.core.context.GlobalContext.startKoin
 import org.matomo.sdk.TrackerBuilder
 import org.matomo.sdk.extra.MatomoApplication
 
-class App : MatomoApplication(){
+class App : MatomoApplication() {
 
     companion object {
+        //todo[SK] remove this, leaking context, use DI
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/App.kt:22]: Again it's OK to add TODOs as long as you plan to address them "soon".

File: app/src/main/java/com/concordium/wallet/App.kt [hunk 5]
```diff
@@ -1,30 +1,40 @@
...
+            "https://concordium.matomo.cloud/matomo.php",
+            5
+        )//todo[SK] move to config
     }
 
     override fun onCreate() {
         super.onCreate()
 
-        Log.d("App starting - setting Log silent if release")
+        Log.d("App starting - setting Log silent if release")//todo[SK] remove. all logs should be cleared before release.
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/App.kt:37]: Looks like a "clear all logs" task?

File: app/src/main/java/com/concordium/wallet/core/config/EnvironmentConfiguration.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.concordium.wallet.core.config
+
+object EnvironmentConfiguration {
+    const val baseUrl = "https://wallet-proxy.mainnet.concordium.software"
+    const val maxTimeout = 10L
+
+    const val SHARED_PREFERENCES_FILE_NAME = "concordiumPref"
+}
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/core/config/EnvironmentConfiguration.kt:8]: Please set up your editor to automatically end the file with a newline.

File: app/src/main/res/values/styles.xml [hunk 7]
```diff
@@ -522,4 +523,36 @@
         <item name="android:background">@color/theme_divider</item>
     </style>
 
+    <!--..........................New Redesign...........................-->
```
Comment (bisgardo) [inline app/src/main/res/values/styles.xml:526]: What is the rationale behind this naming? Is the design actually named "redesign"? Which we're introducing gradually?

File: app/src/main/java/com/concordium/wallet/ui/base/BaseBindingFragment.kt [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import androidx.annotation.LayoutRes
+import androidx.databinding.DataBindingUtil
+import androidx.databinding.ViewDataBinding
+import androidx.fragment.app.Fragment
+import androidx.navigation.fragment.findNavController
+import com.concordium.wallet.util.Log
+
+abstract class BaseBindingFragment<BINDING : ViewDataBinding> : Fragment() {
+
+    protected var _viewDataBinding: BINDING? = null
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/ui/base/BaseBindingFragment.kt:18]: `private`?

File: app/src/main/java/com/concordium/wallet/ui/bakerdelegation/baker/BakerRegistrationConfirmationActivity.kt [hunk 9]
```diff
@@ -46,15 +46,17 @@ class BakerRegistrationConfirmationActivity : BaseDelegationBakerActivity() {
...
     private fun loadFee() {
-        viewModel.transactionFeeLiveData.observe(this, object : Observer<Pair<Long?, Int?>> {
-            override fun onChanged(response: Pair<Long?, Int?>?) {
-                response?.first?.let {
-                    showWaiting(binding.includeProgress.progressLayout, false)
-                    updateViews()
-                    binding.estimatedTransactionFee.text = getString(R.string.delegation_register_delegation_amount_estimated_transaction_fee, CurrencyUtil.formatGTU(it))
-                }
+        viewModel.transactionFeeLiveData.observe(this
+        ) { response ->
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/ui/bakerdelegation/baker/BakerRegistrationConfirmationActivity.kt:50]:
```suggestion
        viewModel.transactionFeeLiveData.observe(this) { response ->
```

Same below.

File: app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    data class UpdateTerms(val newestTerms: TermsAndConditionsItem) : TermsAndConditionsState()
+
+    data class Error(val originalException: Throwable) : TermsAndConditionsState()
+
+    object NavigateOnboardingForward : TermsAndConditionsState()
+    object NavigateUpdateForward : TermsAndConditionsState()
+}
+
+
+
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsViewModel.kt:75]:
```suggestion
```

File: app/build.gradle [hunk 11]
```diff
@@ -339,11 +344,12 @@ dependencies {
     implementation 'androidx.browser:browser:1.5.0'
 
     // OkHttp/Retrofit
-    implementation 'com.squareup.okhttp3:okhttp:4.9.1'
-    implementation 'com.squareup.okhttp3:logging-interceptor:4.9.1'
+    implementation 'com.squareup.okhttp3:okhttp:4.10.0'
+    implementation 'com.squareup.okhttp3:logging-interceptor:4.10.0'
```
Comment (bisgardo) [inline app/build.gradle:348]: Like formatting, unless this kind of change is required for the PR to work, it should be done separately.

It's very important that PRs contain as few unrelated changes as possible. This will get them through review much faster and also make it easier to bisect bugs later on.

File: app/src/main/java/com/concordium/wallet/core/config/EnvironmentConfiguration.kt [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package com.concordium.wallet.core.config
+
+object EnvironmentConfiguration {
+    const val baseUrl = "https://wallet-proxy.mainnet.concordium.software"
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/core/config/EnvironmentConfiguration.kt:4]: Why is this hardcoded to mainnet?

**Review by bisgardo — COMMENTED** [review]
Preliminary review. Please bear with me as I'm not an android expert so it takes a long time to understand all the changes.

If you can split the PR to do formatting changes etc. separately that would be great. But only if it isn't too much trouble.

File: app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+import com.concordium.wallet.ui.base.BaseBindingFragment
+import com.concordium.wallet.ui.intro.introstart.IntroTermsActivity
+import com.google.android.material.snackbar.Snackbar
+import com.walletconnect.util.Empty
+import kotlinx.coroutines.launch
+import org.koin.androidx.viewmodel.ext.android.activityViewModel
+import java.util.regex.Pattern
+
+internal const val TERMS_AND_CONDITIONS_WEB_LINK =
+    "https://developer.concordium.software/en/mainnet/net/resources/terms-and-conditions-bw.html"
```
Comment (orhoj) [inline app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt:21]: This is the old browser wallet terms and conditions. If this variable is used as a fallback, in case that the wallet-proxy does not reply, then it should point to the new location of the terms and conditions:

`https://developer.concordium.software/en/mainnet/net/resources/terms-and-conditions.html`

File: app/build.gradle [hunk 14]
```diff
@@ -16,8 +16,10 @@ tasks.create(name: "unitTestCoverageReport", type: JacocoReport, dependsOn: "tes
     description = "Generate Jacoco coverage reports for the 'local' debug build."
 
     reports {
-        html.enabled = true // human readable (written into './build/reports/jacoco/unitTestCoverageReport/html')
-        xml.enabled = true // CI-readable (written into './build/reports/jacoco/unitTestCoverageReport/unitTestCoverageReport.xml')
+        html.enabled = true
+        // human readable (written into './build/reports/jacoco/unitTestCoverageReport/html')
+        xml.enabled = true
+        // CI-readable (written into './build/reports/jacoco/unitTestCoverageReport/unitTestCoverageReport.xml')
```
[the remark this answers is not in the document]
Comment (czerwix) [reply app/build.gradle:22]: done

File: app/build.gradle [hunk 15]
```diff
@@ -52,25 +54,25 @@ task printVersionName {
 
 //def LOCAL_IP = getLocalIPv4()
 def LOCAL_IP = "10.0.2.2"
-def URL_PROXY_BASE_LOCAL = "http://"+LOCAL_IP+":14000"
+def URL_PROXY_BASE_LOCAL = "http://" + LOCAL_IP + ":14000"
```
[the remark this answers is not in the document]
Comment (czerwix) [reply app/build.gradle:57]: I got an automatic formatter/lint running on all changed files when committing. will turn this off in the future.

File: app/build.gradle [hunk 16]
```diff
@@ -300,7 +303,7 @@ android {
     }
 }
 
-dependencies {
+dependencies {//todo move to proper gradle dependencies management solution.
```
[the remark this answers is not in the document]
Comment (czerwix) [reply app/build.gradle:306]: sure. the reference was to the `dependencies` section. I can remove all the todos, no problem from my side. Just something to keep a note of work needed to be done in the future.

[the remark this answers is not in the document]
Comment (czerwix) [reply app/build.gradle:306]: in Android Studio we have a very nice view of todos. 
image

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (czerwix) [reply app/src/main/AndroidManifest.xml:49]: removed.

File: app/src/main/java/com/concordium/wallet/ui/base/BaseBindingFragment.kt (hunk 8 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/ui/base/BaseBindingFragment.kt:18]: This access should not be restricted as in some cases we need complete control over binding in fragments. The lifecycle bound non-null viewDataBinding is here for convenience and code clarity/quality.

File: app/src/main/java/com/concordium/wallet/ui/bakerdelegation/baker/BakerRegistrationConfirmationActivity.kt (hunk 9 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/ui/bakerdelegation/baker/BakerRegistrationConfirmationActivity.kt:50]: done

File: app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsViewModel.kt (hunk 10 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsViewModel.kt:75]: done

File: app/build.gradle (hunk 11 above)
Comment (czerwix) [reply app/build.gradle:348]: sure.

File: app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt (hunk 13 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt:21]: had the wrong link, updated it,

File: app/src/main/java/com/concordium/wallet/core/config/EnvironmentConfiguration.kt (hunk 12 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/core/config/EnvironmentConfiguration.kt:4]: removed it, for now, and will update the configuration to all flavours with separate PR/task. 
This is base class that will expand when I move environment configuration from gradle->EnvironmentConfiguration  object, which is defined per flavour package.

File: app/build.gradle (hunk 1 above)
Comment (czerwix) [reply app/build.gradle:386]: kk, done

File: app/build.gradle [hunk 17]
```diff
@@ -339,11 +344,12 @@ dependencies {
...
     // OkHttp/Retrofit
-    implementation 'com.squareup.okhttp3:okhttp:4.9.1'
-    implementation 'com.squareup.okhttp3:logging-interceptor:4.9.1'
+    implementation 'com.squareup.okhttp3:okhttp:4.10.0'
+    implementation 'com.squareup.okhttp3:logging-interceptor:4.10.0'
     implementation 'com.squareup.retrofit2:retrofit:2.9.0'
     implementation 'com.squareup.retrofit2:converter-gson:2.9.0'
-    implementation 'com.squareup.retrofit2:converter-scalars:2.6.2'
+    implementation 'com.squareup.retrofit2:converter-scalars:2.9.0'
+    implementation "com.squareup.retrofit2:converter-moshi:2.9.0"
```
Comment (czerwix) [inline app/build.gradle:352]: added mochi for easy json handling.

File: app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (czerwix) [reply app/src/main/AndroidManifest.xml:63]: that is the most common formatting style in Android. This is for better visibility between Actions/Categories etc.

File: app/src/main/java/com/concordium/wallet/App.kt (hunk 4 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/App.kt:22]: yes, for now i have removed all mine todos and added to personal notes, will create tasks for these changes.

File: app/src/main/java/com/concordium/wallet/App.kt (hunk 5 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/App.kt:37]: yes, removed todo

File: app/src/main/java/com/concordium/wallet/core/config/EnvironmentConfiguration.kt (hunk 6 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/core/config/EnvironmentConfiguration.kt:8]: I never understood this approach. could I ask what is significant in the extra line or lack of it at the end of the file?

File: app/src/main/res/values/styles.xml (hunk 7 above)
Comment (czerwix) [reply app/src/main/res/values/styles.xml:526]: yes, the new onboarding flow comes with the redesign of the UI experience. new styles etc will be added and I do not want to mix things up. this will most likely change in future as I get time for proper tasks that refactor the structure here.

**Review by bisgardo — COMMENTED** [review]
I don't know if there's something wrong with my setup, but I still get the wall of text display on a fresh install. That shouldn't happen right?

**Comment by bisgardo** [comment]
Would it be possible to add tests for at least the basic flow? Like https://github.com/Concordium/concordium-reference-wallet-ios/pull/310/files#diff-b0bca1c18e26af94fc0acfe89180d8ade699ea2cd48b36183f93235b02f1ada7.

File: app/build.gradle (hunk 15 above)
[the remark this answers is not in the document]
Comment (czerwix) [reply app/build.gradle:57]: sure, i'll do that. will run the default(most used) formatting for the entire project

File: app/src/main/java/com/concordium/wallet/onboarding/data/model/TermsAndConditionsDto.kt [hunk 18]
```diff
@@ -0,0 +1,6 @@
+package com.concordium.wallet.onboarding.data.model
+
+data class TermsAndConditionsDto(
```
[the remark this answers is not in the document]
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/onboarding/data/model/TermsAndConditionsDto.kt:3]: yes, the DTO is only for transport as name suggest, where as Item is UI specific model that can be different.

File: app/src/main/java/com/concordium/wallet/App.kt [hunk 19]
```diff
@@ -35,6 +44,23 @@ class App : MatomoApplication(){
...
+        startKoin {
+            androidLogger()
+            androidContext(this@App)
+            modules(
+                modules.configModule,
+                modules.sharedPreferencesModule,
+                modules.remoteModule,
+                modules.apiModule,
+            )
+            modules(OnboardingModule.modules)
```
[the remark this answers is not in the document]
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/App.kt:62]: yes. all features will be added to separate modules. so we can load/unload dependencies when necessary?

File: app/build.gradle (hunk 15 above)
[the remark this answers is not in the document]
Comment (bisgardo) [reply app/build.gradle:57]: Just to be clear I meant in a separate PR :)

It would be great if you could do that right away, then we'll revert the formatting changes from this PR (I can do that). This would make it much easier for me to understand the changes.

File: app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+                            viewDataBinding.loadingContainer.visibility = View.VISIBLE
+
+                        is TermsAndConditionsState.InitialTerms -> updateInitialTermsBinding()
+                        is TermsAndConditionsState.UpdateTerms -> updateUpdateTermsBinding()
+                        TermsAndConditionsState.NavigateOnboardingForward -> {
+                            startActivity(Intent(requireContext(), IntroTermsActivity::class.java))
+                        }
+
+                        TermsAndConditionsState.NavigateUpdateForward -> {
+                            //todo[SK] go back to wallet.
```
[the remark this answers is not in the document]
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt]: fixed

**Comment by czerwix** [comment]
> I don't know if there's something wrong with my setup, but I still get the wall of text display on a fresh install. That shouldn't happen right?

yes, fixed now.

File: app/src/main/java/com/concordium/wallet/core/arch/BackendObserver.kt [hunk 21]
```diff
@@ -11,7 +11,7 @@ import com.concordium.wallet.core.backend.BackendResource
 
 class BackendObserver<T> : Observer<BackendResource<T>> {
 
-    override fun onChanged(backendResource: BackendResource<T>?) {
+    override fun onChanged(backendResource: BackendResource<T>) {
         onDone()
         if (backendResource != null) {
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/core/arch/BackendObserver.kt:16]: Redundant check as the var isn't nullable anymore.

File: app/src/main/java/com/concordium/wallet/core/arch/EventObserver.kt [hunk 22]
```diff
@@ -10,7 +10,7 @@ import androidx.lifecycle.Observer
 
 open class EventObserver<T> : Observer<Event<T>> {
 
-    override fun onChanged(event: Event<T>?) {
+    override fun onChanged(event: Event<T>) {
         if (event != null) {
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/core/arch/EventObserver.kt:14]: same

File: app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+import com.concordium.wallet.ui.base.BaseBindingFragment
+import com.concordium.wallet.ui.intro.introstart.IntroStartActivity
+import com.google.android.material.snackbar.Snackbar
+import com.walletconnect.util.Empty
+import kotlinx.coroutines.launch
+import org.koin.androidx.viewmodel.ext.android.activityViewModel
+import java.util.regex.Pattern
+
+internal const val TERMS_AND_CONDITIONS_WEB_LINK =
+    "https://developer.concordium.software/en/mainnet/net/resources/terms-and-conditions.html"
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt:23]: This should not be hardcoded. It should instead use the `url` field from `TermsAndConditionsItem` which, as far as I can tell, isn't currently being used for anything?

Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/onboarding/ui/terms/TermsAndConditionsFragment.kt:23]: fixed

File: app/src/main/java/com/concordium/wallet/core/arch/EventObserver.kt (hunk 22 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/core/arch/EventObserver.kt:14]: fixed

File: app/src/main/java/com/concordium/wallet/core/arch/BackendObserver.kt (hunk 21 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/core/arch/BackendObserver.kt:16]: fixed

File: app/src/main/java/com/concordium/wallet/core/arch/BackendEventObserver.kt [hunk 24]
```diff
@@ -12,24 +12,19 @@ import com.concordium.wallet.core.backend.BackendEventResource
 
 open class BackendEventObserver<T> : Observer<BackendEventResource<T>> {
 
-    override fun onChanged(backendResource: BackendEventResource<T>?) {
+    override fun onChanged(backendResource: BackendEventResource<T>) {
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/core/arch/BackendEventObserver.kt:15]: I can tell that this change is necessary for the code to compile now, but I don't see why. Can you help me understand? What changed for this to be the case?

File: app/src/main/java/com/concordium/wallet/onboarding/data/OnboardingRepository.kt [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+    private val sharedPreferences: SharedPreferencesDataSource,
+    private val onboardingService: OnboardingService
+) {
+
+    fun saveAcceptedTermsAndConditionsVersion(version: String) = runCatching {
+        sharedPreferences.termsAndConditionsVersionAccepted = version
+    }
+
+    fun getLocalAcceptedTermsAndConditionsVersion(): String =
+        sharedPreferences.termsAndConditionsVersionAccepted
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/onboarding/data/OnboardingRepository.kt:17]: I removed `suspend` as Android Studio claimed that they were redundant. Please confirm.

File: app/src/main/java/com/concordium/wallet/onboarding/ui/di/ModuleLoader.kt [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+abstract class ModuleLoader {
+    abstract val modules: List<Module>
+
+    fun load() {
+        loadKoinModules(modules)
+    }
+
+    fun unload() {
+        unloadKoinModules(modules)
+    }
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/onboarding/ui/di/ModuleLoader.kt:16]: According to Studio, none of these methods are used.

File: app/src/main/java/com/concordium/wallet/ui/MainActivity.kt [hunk 27]
```diff
@@ -208,6 +209,14 @@ class MainActivity : BaseActivity(), IdentityStatusDelegate by IdentityStatusDel
             newAccount?.let {
             }
         }
+        viewModel.showTermsAndConditions.observe(this) { shouldShowTermsAndConditions->
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/ui/MainActivity.kt:212]:
```suggestion
        viewModel.showTermsAndConditions.observe(this) { shouldShowTermsAndConditions ->
```

File: app/src/main/java/com/concordium/wallet/ui/recipient/recipientlist/RecipientListViewModel.kt [hunk 28]
```diff
@@ -18,7 +18,7 @@ class RecipientListViewModel(application: Application) : AndroidViewModel(applic
 
     private val allRecipientsLiveData: LiveData<List<Recipient>>
     val recipientListLiveData: LiveData<List<Recipient>>
-        get() = Transformations.switchMap(allRecipientsLiveData) { allRecipients ->
+        get() = allRecipientsLiveData.switchMap { allRecipients ->
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/ui/recipient/recipientlist/RecipientListViewModel.kt:21]: Why was this change necessary?

**Review by bisgardo — APPROVED** [review]
Looks good! The only minor thing I encountered when verifying is that the button says "Create password" even when the password is already set, meaning that it will go to the "enter password" screen. This could be addressed in a followup.

Btw. I really appreciate the screen captures :+1:

File: app/src/main/res/values/strings.xml [hunk 29]
```diff
@@ -996,4 +996,15 @@
...
     <string name="export_account_keys_file_exported">File exported</string>
     <string name="wallet_connect_transaction_parsing_error">Error parsing the request!</string>
+
+    <string name="terms_and_conditions_title">Before you begin</string>
+    <string name="terms_and_conditions_title_update">Updated terms and conditions</string>
+    <string name="terms_and_conditions_description">Before you start using the Concordium Mobile Wallet, you have to set up a passcode and optionally biometrics.\n\nIt is very important to keep your passcode safe, because it is the only way to access your accounts. Concordium is not be able to change your passcode or help you unlock your wallet if you lose your passcode.</string>
+    <string name="terms_and_conditions_description_update">The terms and conditions have been updated. Please review the terms and conditions before continuing.</string>
+
+    <string name="terms_and_conditions_create_password">Create password</string>
+    <string name="terms_and_conditions_create_password_update">Continue</string>
```
[the remark this answers is not in the document]
Comment (czerwix) [reply app/src/main/res/values/strings.xml:1006]: fixed

File: app/src/main/java/com/concordium/wallet/ui/recipient/recipientlist/RecipientListViewModel.kt (hunk 28 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/ui/recipient/recipientlist/RecipientListViewModel.kt:21]: the Transformations.switchMap was deprecated and moved to extension function.

File: app/src/main/java/com/concordium/wallet/core/arch/BackendEventObserver.kt (hunk 24 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/core/arch/BackendEventObserver.kt:15]: the value in super fun is not nullable so it should not be here. 
image

File: app/src/main/java/com/concordium/wallet/onboarding/ui/di/ModuleLoader.kt (hunk 26 above)
Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/onboarding/ui/di/ModuleLoader.kt:16]: we can keep it this way for now, I'll change it when necessary

File: app/src/main/java/com/concordium/wallet/core/arch/BackendEventObserver.kt (hunk 24 above)
Comment (bisgardo) [reply app/src/main/java/com/concordium/wallet/core/arch/BackendEventObserver.kt:15]: But then why did it work before? From what I can tell this interface didn't change?

File: app/src/main/java/com/concordium/wallet/onboarding/data/OnboardingRepository.kt (hunk 25 above)
Comment (bisgardo) [reply app/src/main/java/com/concordium/wallet/onboarding/data/OnboardingRepository.kt:17]: ok feel free to re-add it then

Comment (czerwix) [reply app/src/main/java/com/concordium/wallet/onboarding/data/OnboardingRepository.kt:17]: for the time being, we can leave it like this, I'll add this when necessary.
